service/debugger/debugger.go

```diff
@@ -0,0 +1,519 @@
+	debugger := &Debugger{
+		processOps: make(chan func(*proctl.DebuggedProcess)),
+		config:     config,
+		stop:       make(chan stopSignal),
+	}
+	return debugger
+}
+
+// withProcess facilitates thread-safe access to the DebuggedProcess. Most
+// interaction with DebuggedProcess should occur via calls to withProcess[1],
+// and the functions placed on the processOps channel should be consumed and
+// executed from the same thread as the DebuggedProcess.
+//
+// This is convenient because it allows things like HTTP handlers in
+// goroutines to work with the DebuggedProcess with synchronous semantics.
+//
+// [1] There are some exceptional cases where direct access is okay; for
+// instance, when performing an operation like halt which merely sends a
+// signal to the process rather than performing something like a ptrace
+// operation.
+func (d *Debugger) withProcess(f func(*proctl.DebuggedProcess) error) error {
+	if !d.running {
+		return fmt.Errorf("debugger isn't running")
+	}
+
+	result := make(chan error)
+	d.processOps <- func(proc *proctl.DebuggedProcess) {
+		result <- f(proc)
+	}
+	return <-result
+}
+
+// Run starts debugging a process until Detach is called.
+func (d *Debugger) Run() error {
+	// We must ensure here that we are running on the same thread during
+	// the execution of dbg. This is due to the fact that ptrace(2) expects
+	// all commands after PTRACE_ATTACH to come from the same thread.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	d.running = true
+	defer func() { d.running = false }()
+
+	// Create the process by either attaching or launching.
+	if d.config.AttachPid > 0 {
+		log.Printf("attaching to pid %d", d.config.AttachPid)
+		p, err := proctl.Attach(d.config.AttachPid)
+		if err != nil {
+			return fmt.Errorf("couldn't attach to pid %d: %s", d.config.AttachPid, err)
+		}
+		d.process = p
+	} else {
+		log.Printf("launching process with args: %v", d.config.ProcessArgs)
+		p, err := proctl.Launch(d.config.ProcessArgs)
+		if err != nil {
+			return fmt.Errorf("couldn't launch process: %s", err)
+		}
+		d.process = p
+	}
+
+	// Handle access to the process from the current thread.
+	log.Print("debugger started")
+	for {
+		select {
+		case f := <-d.processOps:
+			// Execute the function
+			f(d.process)
+		case s := <-d.stop:
+			// Handle shutdown
+			log.Print("debugger is stopping")
+
+			// Clear breakpoints
+			bps := []*proctl.BreakPoint{}
+			for _, bp := range d.process.BreakPoints {
+				if bp != nil {
+					bps = append(bps, bp)
+				}
+			}
+			for _, bp := range d.process.HWBreakPoints {
+				if bp != nil {
+					bps = append(bps, bp)
+				}
+			}
+			for _, bp := range bps {
+				_, err := d.process.Clear(bp.Addr)
+				if err != nil {
+					log.Printf("warning: couldn't clear breakpoint @ %#v: %s", bp.Addr, err)
+				} else {
+					log.Printf("cleared breakpoint @ %#v", bp.Addr)
+				}
+			}
+
+			// Detach
+			if err := sys.PtraceDetach(d.process.Pid); err == nil {
```

Comment:
The process could still be running during this shutdown sequence, which means that `exited` could be changing concurrently, correct? If so, without a lock, there's a race where we could check `exited`, see `true`, and by the time we detach, the process has actually exited. Agree?

Is there a specific error type I could check so that I could ignore attempted detach calls against an exited proc?

In any event, any errors (benign or not) during shutdown are merely logged, so I don't know that it's a huge deal either way. Let me know your final thoughts on this piece and I'll defer to your opinion.